compiler/rustc_mir_transform/src/check_pointers.rs

```diff
@@ -0,0 +1,198 @@
+            let statement = &basic_blocks[block].statements[statement_index];
+            let source_info = statement.source_info;
+
+            let mut finder = PointerFinder::new(tcx, local_decls, typing_env, excluded_pointees);
+            finder.visit_statement(statement, location);
+
+            for (local, ty) in finder.into_found_pointers() {
+                debug!("Inserting check for {:?}", ty);
+                let new_block = split_block(basic_blocks, location);
+
+                // Invoke `on_finding` which appends to `local_decls` and the
+                // blocks statements. It returns information about the assert
+                // we're performing in the Terminator.
+                let block_data = &mut basic_blocks[block];
+                let pointer_check = on_finding(
+                    tcx,
+                    local,
+                    ty,
+                    local_decls,
+                    &mut block_data.statements,
+                    source_info,
+                );
+                block_data.terminator = Some(Terminator {
+                    source_info,
+                    kind: TerminatorKind::Assert {
+                        cond: pointer_check.cond,
+                        expected: true,
+                        target: new_block,
+                        msg: pointer_check.assert_kind,
+                        // This calls a panic function associated with the pointer check, which
+                        // is #[rustc_nounwind]. We never want to insert an unwind into unsafe
+                        // code, because unwinding could make a failing UB check turn into much
+                        // worse UB when we start unwinding.
+                        unwind: UnwindAction::Unreachable,
+                    },
+                });
+            }
+        }
+    }
+}
+
+struct PointerFinder<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    local_decls: &'a mut LocalDecls<'tcx>,
+    typing_env: ty::TypingEnv<'tcx>,
+    pointers: Vec<(Place<'tcx>, Ty<'tcx>)>,
+    excluded_pointees: &'a [Ty<'tcx>],
+}
+
+impl<'a, 'tcx> PointerFinder<'a, 'tcx> {
+    fn new(
+        tcx: TyCtxt<'tcx>,
+        local_decls: &'a mut LocalDecls<'tcx>,
+        typing_env: ty::TypingEnv<'tcx>,
+        excluded_pointees: &'a [Ty<'tcx>],
+    ) -> Self {
+        PointerFinder { tcx, local_decls, typing_env, excluded_pointees, pointers: Vec::new() }
+    }
+
+    fn into_found_pointers(self) -> Vec<(Place<'tcx>, Ty<'tcx>)> {
+        self.pointers
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for PointerFinder<'a, 'tcx> {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
+        // We want to only check reads and writes to Places, so we specifically exclude
+        // Borrow and RawBorrow.
+        match context {
+            PlaceContext::MutatingUse(
+                MutatingUseContext::Store
+                | MutatingUseContext::AsmOutput
```

Comment:
Does AsmOutput really require alignment and non-null? (I don't entirely know what it means to be an "asm output operand".)

Reply:
Judging from the [visitor code](https://github.com/rust-lang/rust/blob/master/compiler/rustc_middle/src/mir/visit.rs#L593-L609) this seems to comes only from inline assembly blocks and their operands (for `out` and `inout` operands). I personally believe this makes sense to also check, but I am happy to be convinced otherwise. Anyways if we believe to keep this we should add a test for it and also for the alignment case.

@saethlin who intially added this `PlaceContext`, would be interesting to get your opinion here!

Reply:
Oke I tried to write a test and turns out the current implementation doesn't even sanitize `AsmOutput` places (at least I didn't find code that triggers that). So AFAIC we could leave this away without any effect. 

My hypothesis is that `AsmOutput` places can only occur for `TerminatorKind::InlineAsm` which our current `PointerFinder` doesn't visit since it only looks at a blocks statement.

Maybe we should first decide whether we want to sanitize this or not before investing into implementing it.